commit f1f4571fb52398a27257a116657cb49bfd143e1b
Author: AdarshGrewal <adarshgrewal@gmail.com>
Date:   Sat Aug 10 14:08:50 2024 +0530

    fix fp

diff --git a/packages/SystemUI/Android.bp b/packages/SystemUI/Android.bp
index dd5c9d0fac87c..3087d9dd26a70 100644
--- a/packages/SystemUI/Android.bp
+++ b/packages/SystemUI/Android.bp
@@ -526,6 +526,8 @@ android_library {
         "device_policy_aconfig_flags_lib",
         "vendor.lineage.powershare-V1.0-java",
         "FadingEdgeLayoutLib",
+        "vendor.xiaomi.hardware.fingerprintextension-V1-java",
+        "vendor.xiaomi.hw.touchfeature-V1-java"
     ],
     libs: [
         "keepanno-annotations",
@@ -756,7 +758,7 @@ android_app {
     defaults: [
         "platform_app_defaults",
         "SystemUI_optimized_defaults",
-    ],
+   ],
     manifest: "tests/AndroidManifest-base.xml",
 
     srcs: [
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
index bf89cbb78937b..ff38c0ddf0765 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
@@ -135,6 +135,12 @@ import java.util.concurrent.Executor;
 
 import javax.inject.Inject;
 
+import kotlinx.coroutines.CoroutineScope;
+import kotlinx.coroutines.ExperimentalCoroutinesApi;
+
+import vendor.xiaomi.hardware.fingerprintextension.IXiaomiFingerprint;
+import vendor.xiaomi.hw.touchfeature.ITouchFeature;
+
 /**
  * Shows and hides the under-display fingerprint sensor (UDFPS) overlay, handles UDFPS touch events,
  * and toggles the UDFPS display mode.
@@ -199,6 +205,9 @@ public class UdfpsController implements DozeReceiver, Dumpable {
     private final boolean mIgnoreRefreshRate;
     private final KeyguardTransitionInteractor mKeyguardTransitionInteractor;
 
+    private static ITouchFeature xaiomiTouchFeatureAidl = null;
+    private static IXiaomiFingerprint xaiomiFingerprintExtensionAidl = null;
+
     // Currently the UdfpsController supports a single UDFPS sensor. If devices have multiple
     // sensors, this, in addition to a lot of the code here, will be updated.
     @VisibleForTesting @NonNull FingerprintSensorPropertiesInternal mSensorProps;
@@ -299,6 +308,49 @@ public class UdfpsController implements DozeReceiver, Dumpable {
                     updateUdfpsAnimation();
                 }
             };
+    private static void xaiomiTouchFeature(int arg) {
+        try {
+            if (xaiomiTouchFeatureAidl == null) {
+                var name = "default";
+                var fqName = vendor.xiaomi.hw.touchfeature.ITouchFeature.DESCRIPTOR + "/" + name;
+                var b = android.os.Binder.allowBlocking(android.os.ServiceManager.waitForDeclaredService(fqName));
+                xaiomiTouchFeatureAidl = vendor.xiaomi.hw.touchfeature.ITouchFeature.Stub.asInterface(b);
+                
+                // Link to death
+                b.linkToDeath(() -> {
+                    android.util.Log.w("FP-HAX", "TouchFeature binder died. Reconnecting...");
+                    xaiomiTouchFeatureAidl = null;
+                }, 0);
+                
+                android.util.Log.d("FP-HAX", "Binded TouchFeature");
+            }
+            xaiomiTouchFeatureAidl.setTouchMode(0, 10, arg);
+        } catch(Throwable t) {
+            android.util.Log.e("FP-HAX", "TouchFeature", t);
+        }
+    }
+
+    private static void xiaomiFingerprintExtension(int arg) {
+        try {
+            if (xaiomiFingerprintExtensionAidl == null) {
+                var name = "default";
+                var fqName = vendor.xiaomi.hardware.fingerprintextension.IXiaomiFingerprint.DESCRIPTOR + "/" + name;
+                var b = android.os.Binder.allowBlocking(android.os.ServiceManager.waitForDeclaredService(fqName));
+                xaiomiFingerprintExtensionAidl = vendor.xiaomi.hardware.fingerprintextension.IXiaomiFingerprint.Stub.asInterface(b);
+                
+                // Link to death
+                b.linkToDeath(() -> {
+                    android.util.Log.w("FP-HAX", "FingerprintExtension binder died. Reconnecting...");
+                    xaiomiFingerprintExtensionAidl = null;
+                }, 0);
+                
+                android.util.Log.d("FP-HAX", "Binded FingerprintExtension");
+            }
+            xaiomiFingerprintExtensionAidl.extCmd(4, arg);
+        } catch(Throwable t) {
+            android.util.Log.e("FP-HAX", "FingerprintExtension", t);
+        }
+    }
 
     @Override
     public void dump(@NonNull PrintWriter pw, @NonNull String[] args) {
@@ -1269,6 +1321,8 @@ public class UdfpsController implements DozeReceiver, Dumpable {
                 }
             }
         }
+        xaiomiTouchFeature(1);
+        xiaomiFingerprintExtension(1);
 
         for (Callback cb : mCallbacks) {
             cb.onFingerDown();
@@ -1341,6 +1395,24 @@ public class UdfpsController implements DozeReceiver, Dumpable {
     private void hideUdfpsAnimation() {
         if (mUdfpsAnimation != null) {
             mUdfpsAnimation.hide();
+        xaiomiTouchFeature(0);
+        xiaomiFingerprintExtension(0);
+
+        // Add a delay to ensure that the dim amount is updated after the display has had chance
+        // to switch out of HBM mode. The delay, in ms is stored in config_udfpsDimmingDisableDelay.
+        // If the delay is 0, the dim amount will be updated immediately.
+        final int delay = mContext.getResources().getInteger(
+                com.android.systemui.res.R.integer.config_udfpsDimmingDisableDelay);
+        if (delay > 0) {
+            mFgExecutor.executeDelayed(() -> {
+                // A race condition exists where the overlay is destroyed before the dim amount
+                // is updated. This check ensures that the overlay is still valid.
+                if (mOverlay != null && mOverlay.matchesRequestId(requestId)) {
+                    updateViewDimAmount();
+                }
+            }, delay);
+        } else {
+            updateViewDimAmount();
         }
     }
 
diff --git a/services/core/Android.bp b/services/core/Android.bp
index 38319006d103c..b0f7c46b7e236 100644
--- a/services/core/Android.bp
+++ b/services/core/Android.bp
@@ -231,6 +231,8 @@ java_library_static {
         "aconfig_new_storage_flags_lib",
         "powerstats_flags_lib",
         "vendor.aospa.biometrics.face",
+        "vendor.xiaomi.hardware.fingerprintextension-V1-java",
+        "vendor.xiaomi.hw.touchfeature-V1-java",
     ],
     javac_shard_size: 50,
     javacflags: [
